hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java

```diff
@@ -0,0 +1,446 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership.  The ASF
+ * licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.hadoop.ozone.om;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.hdds.utils.db.Table;
+import org.apache.hadoop.hdds.utils.db.TableIterator;
+import org.apache.hadoop.ozone.om.helpers.SnapshotInfo;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.LinkedHashMap;
+import java.util.TreeMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is used for creating and accessing Snapshot Chains.
+ *
+ * The snapshot chain maintains the in-memory sequence of snapshots
+ * created in chronological order.  There are two such snapshots maintained
+ * i.) Path based snapshot chain, sequence of snapshots created for a
+ * given /volume/bucket
+ * ii.) Global snapshot chain, sequence of all snapshots created in order
+ *
+ * On start, the snapshot chains are initialized from the on disk
+ * SnapshotInfoTable from the om RocksDB.
+ */
+public class SnapshotChainManager {
+  private LinkedHashMap<String, SnapshotChainInfo>  snapshotChainGlobal;
```

Comment:
Some of the structures are defined as "LinkedHashMap"s instead of just HashMaps.  I know that those preserve the insertion order, but that feature doesn't seem used anywhere.  

Instead, a separate doubly linked list structure is implemented through the SnapshotChainInfo prev/next fields?  Given that, are the LinkedHashMap's still needed?

Reply:
The `LinkedHashMap `preserves the order of insertion for little or no cost over the HashMap.  In using the LinkedHashMap we can easily access an iterator to iterate through the data structure in-order (order is necessary for the chain), something not preserved with the HashMap.  The `LinkedHashMap` also gives us constant time inserts, deletes and lookups.